### ISSUE TITLE
Very low mAP afte training with over 7K imgs while applying pretrained weight

### ISSUE THREAD

**Issue by NosremeC** [ultralytics/yolov5#699]
## ❔Question
I wanted to train a model which could detect 14 classes. I had over 7K labelled images on hand. The training started last night with the pretrained weight yolov5m.pt. Here are my results after 172 epochs. 

As you can see the mAP is super low. I used to train the same model with same amount of pics on yolov3 and it gave me a better mAP than I have got here. I checked all the labels and them seemed to be fine. I had tried almost all the models (s/m/l/x) and none of them could give me a decent result. I had also looked up the other issues with similar problems but I could not find a reason (maybe it has something to do with the auto-anchor?)

Did I make some silly mistakes here? I have been stuck on this issue for a while and will be so appreciated if anyone can give me some hints. Thanks!

Colab screenshot:
![Colab](https://user-images.githubusercontent.com/68608242/89805985-70f9e980-db04-11ea-9628-909bfe5cfda4.PNG)

results.png:
![results](https://user-images.githubusercontent.com/68608242/89805991-73f4da00-db04-11ea-89da-9dc25f5a35cd.png)

train_batch1.jpg:
![train_batch1](https://user-images.githubusercontent.com/68608242/89806000-77886100-db04-11ea-91c5-df79d11d0573.jpg)

test_batch0_gt.jpg:
![test_batch0_gt](https://user-images.githubusercontent.com/68608242/89806006-79eabb00-db04-11ea-9fdb-c69daa737910.jpg)

test_batch0_pred.jpg:
![test_batch0_pred](https://user-images.githubusercontent.com/68608242/89806016-7c4d1500-db04-11ea-9e51-8f784b5435bb.jpg)

**Comment by glenn-jocher** [comment]
@NosremeC images and curves appear normal, mAP is low. You mentioned autoanchor, can you show a screenshot during autoanchor computation?

**Comment by glenn-jocher** [comment]
@NosremeC I looked at your command, it shows --adam and 173 epochs. 

Before doing anything else or asking anymore questions train using **default settings**.

**Comment by NosremeC** [comment]
> @NosremeC images and curves appear normal, mAP is low. You mentioned autoanchor, can you show a screenshot during autoanchor computation?

Hi, thanks for the advice. I have not actually looked into the autoanchor too much yet, I know there is a parameter called 'noautoanchor' which I just left it as default when training it, I know it might sound silly but what do you mean by 'screenshot during autoanchor computation' specifically?

Several days ago I actually trained a model with yolov5m from scratch with the same dataset up to epoch 290. I have the result here but **the valid dataset used was the same as the train dataset**. Do you think the mAP is reasonable in this case?
yolov5m_epoch290:
![results_290](https://user-images.githubusercontent.com/68608242/89820808-90504100-db1b-11ea-8f23-b69f302cf74f.png)

Also I have noticed one thing when I review the hyps. In your code it stated the default lr for adam is 1e-3, but I carefully walked through train.py and i did not see where it was set that way. It was 1e-2 for both SGD and adam, I put the screenshot here. Could this possibly be an issue?
![lr0](https://user-images.githubusercontent.com/68608242/89820837-9b0ad600-db1b-11ea-90b8-7b691c950a17.PNG)

**Comment by glenn-jocher** [comment]
@NosremeC Autoanchor computes your BPR first, and then solves for new anchors if your BPR is too low (<0.98). If not, the default anchors are fine, no action is taken.

**Forget adam**, you want to train with the default settings first. If you don't see acceptable performance there then something is wrong in your training pipeline, or your dataset is simply a difficult dataset.

<img width="996" alt="Screen Shot 2020-08-10 at 12 13 07 PM" src="https://user-images.githubusercontent.com/26833433/89821155-072d1000-db03-11ea-80d9-940d64c876f1.png">

**Comment by NosremeC** [comment]
> @NosremeC Autoanchor computes your BPR first, and then solves for new anchors if your BPR is too low (<0.98). If not, the default anchors are fine, no action is taken.
> 
> **Forget adam**, you want to train with the default settings first. If you don't see acceptable performance there then something is wrong in your training pipeline, or your dataset is simply a difficult dataset.
> 
> <img alt="Screen Shot 2020-08-10 at 12 13 07 PM" width="996" src="https://user-images.githubusercontent.com/26833433/89821155-072d1000-db03-11ea-80d9-940d64c876f1.png">

Thanks a lot man! I am retraining the model using the latest clone from scratch, I will run for 300 epochs as the default settings. I do not think it is a difficult dataset since the model trained on yolov3 could give me a somewhat nice result, but I really want to have a model from yolov5. I will post some updates here once I finish. Thank you for the patience!

**Comment by glenn-jocher** [comment]
@NosremeC yes exactly, just start from a clean slate, new git clone, let it autodownload the pretrained model again, all default settings, etc. 

But you should screenshot the autoanchor portion to see what it's doing. I'm not aware of any issues with it, but since this is a new product it's possible it may still have implementation issues.

**Comment by NosremeC** [comment]
> @NosremeC yes exactly, just start from a clean slate, new git clone, let it autodownload the pretrained model again, all default settings, etc.
> 
> But you should screenshot the autoanchor portion to see what it's doing. I'm not aware of any issues with it, but since this is a new product it's possible it may still have implementation issues.

This is the autocnhor when I trained my yolov5m model last night:
![autoanchor](https://user-images.githubusercontent.com/68608242/89825669-ed032a00-db22-11ea-9332-2d6d53739b55.PNG)

I am now training a yolov5s model so I can have the completed model before the night and test it. So far I have run around 30 epochs and the result looks much more reasonable than before but I have no idea what caused it. I will run more tests after I have this model.
![s_autoanchor](https://user-images.githubusercontent.com/68608242/89825961-6dc22600-db23-11ea-9021-e3276655c45a.PNG)

**Comment by glenn-jocher** [comment]
@NosremeC right. Autoanchor takes no action in your case, it's a non-factor in your results.

**Comment by NosremeC** [comment]
> @NosremeC right. Autoanchor takes no action in your case, it's a non-factor in your results.

Hey, I have trained a 5s model with 300 epochs(default settings, started with yolov5s.pt), here is the result:

![results_5s](https://user-images.githubusercontent.com/68608242/89930604-036bbd00-dbd9-11ea-8581-6c3e82478cb9.png)

The accuracy of using 5s model on my testset( includes around 1K pics) was about 0.45.

Also I trained a 5m model for about 140 epochs(started with pretrained weight yolov5m.pt), the mAP went down after some point so I stopped the training:

![results_5m](https://user-images.githubusercontent.com/68608242/89930946-83922280-dbd9-11ea-9139-378e50772030.png)

Do they look reasonable or maybe I should try to change the way of training it?

**Comment by glenn-jocher** [comment]
@NosremeC I don't see any obvious problems! I would say the issue title "very low mAP" no longer applies 😃 .

If you want better results, the paths are clear and always the same:
- higher resolution
- larger model
- more data
- evolve hyps

**Comment by NosremeC** [comment]
> @NosremeC I don't see any obvious problems! I would say the issue title "very low mAP" no longer applies 😃 .
> 
> If you want better results, the paths are clear and always the same:
> 
> * higher resolution
> * larger model
> * more data
> * evolve hyps

Thanks for the suggestions! I am just wondering since the pretrained weight (e.g. yolov5m.pt) is trained on COCO dataset, will it improve anything at all if I simply add more pics from COCO to my own dataset and train the model starting from yolov5m.pt? I believe whatever pics I can get from COCO have already been trained to generate yolov5(s/m/l/x).pt?

**Comment by glenn-jocher** [comment]
@NosremeC I don't know. If all of your classes are also COCO classes, don't bother training anything, just use the official models. If some of them intersect you have to experiment and see what works for you.

**Comment by NosremeC** [comment]
> @NosremeC I don't know. If all of your classes are also COCO classes, don't bother training anything, just use the official models. If some of them intersect you have to experiment and see what works for you.

HI, I have trained several models with SGD, but I would like to try adam. When I added --adam to the command, the lr was still showing 0.01. Should I modify the hyp.yaml to change the lr to 0.001 or it will be auto-scheduled to that number?
